File: app/src/main/java/de/logerbyte/moneyminder/dialogs/editDialog/EditDialogViewModel.java
package de.logerbyte.moneyminder.dialogs.editDialog;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;

import de.logerbyte.moneyminder.db.AppDatabaseManager;
import de.logerbyte.moneyminder.db.expense.Expense;
import de.logerbyte.moneyminder.screens.cashsummary.cashadapter.ViewInterface;
import de.logerbyte.moneyminder.util.DigitUtil;
import de.logerbyte.moneyminder.viewModels.CashViewModel;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class EditDialogViewModel extends AndroidViewModel implements BseDialogListenerViewModel {

    private final BaseDialogListenerView callback;
    private AppDatabaseManager appDatabaseManager;
    CashViewModel cashViewModel;
    ViewInterface viewInterface;
    private BaseDialogListenerView baseDialogListenerView;

    public EditDialogViewModel(Application application, BaseDialogListenerView listener) {
        super(application);
        appDatabaseManager = new AppDatabaseManager(application.getBaseContext());
        callback = listener;
    }

    public void setViewInterface(ViewInterface viewInterface) {
        this.viewInterface = viewInterface;
    }

    public void setCashViewModel(CashViewModel cashViewModel) {
        this.cashViewModel = cashViewModel;
    }

    @Override
    public void editCash() {
        Expense expenseToUpdate = new Expense(cashViewModel.getEntryId(),
                cashViewModel.getCashName().get(), cashViewModel.getCashCategory(),
                cashViewModel.getCashDate().get(),
                Double.valueOf(DigitUtil.commaToDot(cashViewModel.getCashAmount().get())));

        baseDialogListenerView.dismissDialog();

        appDatabaseManager.updateCashItem(expenseToUpdate).subscribeOn(
                Schedulers.computation()).observeOn(AndroidSchedulers.mainThread()).subscribe(
                aBoolean -> viewInterface.onUpdateItem());
    }

    @Override
    public void editCashCancel() {
        callback.dismissDialog();
    }

    public void setDialogViewListener(BaseDialogListenerView baseDialogListenerView) {
        this.baseDialogListenerView = baseDialogListenerView;
    }

}

File: app/src/main/java/de/logerbyte/moneyminder/presentation/dialog/dialogEdit/EditDialogViewModel.kt
package de.logerbyte.moneyminder.presentation.dialog.dialogEdit

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import de.logerbyte.moneyminder.entities.data.viewData.CashViewItem
import de.logerbyte.moneyminder.presentation.cashadapter.AdapterCallBack
import de.logerbyte.moneyminder.framework.database.ExpenseRepo
import de.logerbyte.moneyminder.presentation.dialog.BaseDialogViewListener
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

//class EditDialogViewModel(
//    val expenseRepo: ExpenseRepo,
//    dialogCallback: DialogCallback,
//    val context: Context?,
//    val editDialogCallback: EditDialogCallback) : BaseDialogViewModel1(dialogCallback) {

class EditDialogViewModel @Inject constructor(val expenseRepo: ExpenseRepo): ViewModel(), BaseDialogViewListener {

    private lateinit var adapterCallBack: AdapterCallBack
    private lateinit var data: CashViewItem
    val shallDialogClose = MutableLiveData<Boolean>()
    private val _shallDialogClose = shallDialogClose


    private var _categories = MutableLiveData<ArrayList<String>>()
    val categories: LiveData<ArrayList<String>> = _categories

    init {
        loadCategories()
    }

    fun setData(cashViewViewItem: CashViewItem) {
        this.data = cashViewViewItem
    }

    private fun loadCategories() {
        expenseRepo.categories
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { categories -> _categories.value = categories as ArrayList }
    }

//    override fun onClickOk(view: View) {
//        val expenseToUpdate = Expense(editDialogViewModel.cashViewItem.entryId,
//                editDialogViewModel.cashViewItem.cashName.get(), editDialogViewModel.cashViewItem.cashCategory.get(),
//                editDialogViewModel.cashViewItem.cashDate.get(),
//                Double.valueOf(DigitUtil.commaToDot(editDialogViewModel.cashViewItem.cashAmount.get())),
//                editDialogViewModel.cashViewItem.cashPerson.get()
//            )
//
//        expenseRepo.updateCashItem(expenseToUpdate)
//                .subscribeOn(Schedulers.computation())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe({ aBoolean -> adapterCallBack.onUpdateItem() })
//        super.onClickOk(view)
//    }

    fun setAdapter(adapterCallback: AdapterCallBack) {
        adapterCallBack = adapterCallback
    }

    override fun onClickOk() {
        _shallDialogClose.value = true
    }

    override fun onClickCancel() {
        _shallDialogClose.value = true    }
}
